File: src/lib/telemetry.ts
import { invoke } from "@tauri-apps/api/core";
import { getVersion } from "@tauri-apps/api/app";
import type * as Sentry from "@sentry/browser";

/**
 * Opt-in crash reporting (docs/superpowers/specs/2026-07-23-crash-reporting-design.md).
 * Consent is Rust-owned (see src-tauri/src/telemetry.rs) -- the frontend never
 * touches the consent file directly and does not mirror it into localStorage.
 */

export type ConsentValue = "unset" | "granted" | "denied";

export interface TelemetryConsentInfo {
  consent: ConsentValue;
  /** Captured at Rust startup: whether the Rust SDK actually initialized this
   *  session. Gates the JS SDK too, so a mid-session grant can't start
   *  reporting while the Rust SDK stays inactive (design doc, "Frontend side"). */
  active_this_session: boolean;
}

export async function getTelemetryConsent(): Promise<TelemetryConsentInfo> {
  return invoke<TelemetryConsentInfo>("get_telemetry_consent");
}

export async function setTelemetryConsent(consent: ConsentValue): Promise<void> {
  await invoke("set_telemetry_consent", { consent });
}

// --- path scrubbing -----------------------------------------------------
//
// Mirrors src-tauri/src/telemetry.rs::scrub_paths -- keep both in sync. The
// spec does not pin down the exact replacement token (here "~"), only that
// the username/home directory must not survive and that messages are "kept
// but path-scrubbed" (not dropped).

const HOME_PATH_RE =
  /(?:\/Users\/|\/home\/|\/Volumes\/|\/root\b|[A-Za-z]:[\\/]Users[\\/]|\\\\[^\\/\s]+\\[^\\/\s]+)[^\s]*/g;
const UNSCRUBBED_MARKER_RE =
  /(?:\/Users\/|\/home\/|\/Volumes\/|\/root\b|[A-Za-z]:[\\/]Users[\\/]|\\\\[^\\/\s]+\\[^\\/\s]+)/;
const PROJECT_MARKERS = new Set(["agentpanel", "AgentPanel", "src-tauri", "src"]);

/** How many leading segments (already split on the token's separator) make up
 *  the home/mount prefix to fold into "~". Mirrors telemetry.rs's
 *  `home_prefix_fold_len`. */
function homePrefixFoldLen(segments: string[]): number | null {
  const idx = segments.findIndex((s) => s === "Users" || s === "home" || s === "Volumes");
  if (idx !== -1) return Math.min(idx + 2, segments.length);

  const rootIdx = segments.indexOf("root");
  if (rootIdx !== -1) return rootIdx + 1;

  if (segments.length >= 4 && segments[0] === "" && segments[1] === "") {
    // Leading "\\server\share" splits (on "\\") into ["", "", server, share, ...].
    return 4;
  }
  return null;
}

function redactPathToken(token: string): string {
  const sep = token.includes("\\") ? "\\" : "/";
  const segments = token.split(sep);

  const markerIdx = segments.findIndex((s) => PROJECT_MARKERS.has(s));
  if (markerIdx !== -1) {
    return segments.slice(markerIdx).join(sep);
  }

  const foldLen = homePrefixFoldLen(segments);
  if (foldLen !== null) {
    const rest = segments.slice(foldLen);
    return rest.length === 0 ? "~" : `~${sep}${rest.join(sep)}`;
  }

  return token;
}

/**
 * Redact home-directory/mount-point segments from Windows-, macOS- and
 * Linux-style absolute paths appearing anywhere in `input`. Non-path text is
 * left intact. If, after redaction, one of the recognized prefixes still
 * survives (i.e. the text can't be scrubbed with confidence), it is
 * truncated at that point rather than sent with a username/host attached.
 */
export function scrubPaths(input: string): string {
  const out = input.replace(HOME_PATH_RE, redactPathToken);
  const m = out.match(UNSCRUBBED_MARKER_RE);
  return m && m.index !== undefined ? out.slice(0, m.index) : out;
}

function scrubEvent(event: Sentry.ErrorEvent): Sentry.ErrorEvent {
  if (event.message) event.message = scrubPaths(event.message);
  for (const exception of event.exception?.values ?? []) {
    if (exception.value) exception.value = scrubPaths(exception.value);
    for (const frame of exception.stacktrace?.frames ?? []) {
      if (frame.filename) frame.filename = scrubPaths(frame.filename);
    }
  }
  // Defense-in-depth: the Breadcrumbs integration is dropped entirely below
  // (no automatic breadcrumb source runs), but scrub any that show up anyway
  // (e.g. a future manually-added breadcrumb) rather than assume none ever will.
  for (const crumb of event.breadcrumbs ?? []) {
    if (crumb.message) crumb.message = scrubPaths(crumb.message);
    if (crumb.data) {
      for (const key of Object.keys(crumb.data)) {
        const value = crumb.data[key];
        if (typeof value === "string") crumb.data[key] = scrubPaths(value);
      }
    }
  }
  return event;
}

let sdk: typeof Sentry | null = null;
let initPromise: Promise<void> | null = null;

/**
 * Initialize the JS Sentry SDK, gated on the Rust SDK actually having
 * initialized this session (`active_this_session`) -- never on the raw
 * consent file value (see `TelemetryConsentInfo`). Envelopes are handed to
 * the `sentry-tauri` plugin transport, which routes them to the Rust SDK
 * over Tauri IPC; nothing is sent directly from the WebView.
 *
 * `@sentry/browser` and the plugin's JS helpers are loaded via dynamic
 * `import()` only once `active_this_session` is confirmed -- a
 * never-consented user never pays to bundle/parse them.
 *
 * Never throws -- telemetry failing to start must not block the app.
 */
export async function initTelemetry(): Promise<void> {
  if (initPromise) return initPromise;
  initPromise = (async () => {
    let info: TelemetryConsentInfo;
    try {
      info = await getTelemetryConsent();
    } catch {
      return;
    }
    if (!info.active_this_session) return;

    const [SentryModule, pluginApi] = await Promise.all([
      import("@sentry/browser"),
      import("tauri-plugin-sentry-api"),
    ]);
    sdk = SentryModule;

    let release: string | undefined;
    try {
      release = `agentpanel@${await getVersion()}`;
    } catch {
      release = undefined;
    }

    sdk.init({
      ...pluginApi.defaultOptions,
      release,
      // Every automatic breadcrumb source (DOM clicks, console, fetch/xhr,
      // history navigation, Sentry's own internal breadcrumbs) is dropped
      // outright -- the spec allowlist is "exception type/message/stack, app
      // version, OS/arch" only. DOM breadcrumbs in particular serialize
      // element title/aria-label attributes, and this app's TabBar/Sidebar
      // put absolute cwd/repo paths, commit messages and command lines in
      // those attributes.
      integrations: (integrations) =>
        integrations.filter((i) => i.name !== "BrowserSession" && i.name !== "Breadcrumbs"),
      beforeBreadcrumb: undefined,
      beforeSend: scrubEvent,
    });
  })();
  return initPromise;
}

/**
 * Reports an error to Sentry if (and only if) the SDK has been loaded (see
 * `initTelemetry`) -- a silent no-op otherwise, so callers (e.g.
 * `PaneErrorBoundary`) never need to know whether telemetry is active.
 * Deliberately takes no extra context (e.g. a component stack): that's
 * outside the spec allowlist ("exception type/message/stack, version,
 * OS/arch" only), so there's nothing to scrub-or-drop -- it's just never
 * attached.
 */
export function captureError(error: Error): void {
  sdk?.captureException(error);
}

File: src-tauri/src/telemetry.rs
//! Telemetry consent persistence and privacy scrubbing for the opt-in Sentry
//! crash-reporting feature.
//!
//! Consent lives in a small Rust-owned JSON file (sibling of `store.rs`'s
//! data) in the app config directory, e.g. `telemetry.json`:
//! `{ "consent": "unset" | "granted" | "denied" }`. `main.rs` reads this file
//! synchronously, before `tauri::Builder` runs; missing or unparsable file
//! means `Unset`, which means no telemetry (fail closed). Sentry initializes
//! only when consent is `Granted`; a change of consent takes effect on next
//! launch.
//!
//! `scrub_paths` redacts absolute filesystem paths (which leak OS usernames
//! via the home directory) from exception values/messages before they are
//! sent, for both the Rust and TypeScript/JS Sentry SDKs.
//!
//! See docs/superpowers/specs/2026-07-23-crash-reporting-design.md.

use std::fs;
use std::path::{Path, PathBuf};
use std::sync::OnceLock;

use regex::Regex;
use serde::{Deserialize, Serialize};

/// Stored/transmitted as a lowercase string: `"unset"`, `"granted"`, `"denied"`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Consent {
    Unset,
    Granted,
    Denied,
}

/// On-disk shape of `telemetry.json`.
#[derive(Serialize, Deserialize)]
struct StoredConsent {
    consent: Consent,
}

/// Bundle identifier from `tauri.conf.json`'s `identifier` field -- the same
/// value Tauri's own `app_data_dir()` joins onto the platform data dir.
const BUNDLE_ID: &str = "com.jason.agentpanel";

/// Resolve the real `telemetry.json` location. No `AppHandle` exists yet at
/// the point this is needed (before `tauri::Builder` runs), so this
/// recomputes the same directory Tauri's `app_data_dir()` would resolve to,
/// rather than going through Tauri.
pub fn consent_path() -> PathBuf {
    platform_data_dir().join(BUNDLE_ID).join("telemetry.json")
}

#[cfg(target_os = "windows")]
fn platform_data_dir() -> PathBuf {
    std::env::var_os("APPDATA")
        .map(PathBuf::from)
        .unwrap_or_else(|| PathBuf::from("."))
}

#[cfg(target_os = "macos")]
fn platform_data_dir() -> PathBuf {
    std::env::var_os("HOME")
        .map(|home| PathBuf::from(home).join("Library/Application Support"))
        .unwrap_or_else(|| PathBuf::from("."))
}

// minimal: XDG fallback only; Linux isn't a shipped bundle target today (see
// tauri.conf.json's bundle.targets), but this keeps dev builds/CI on Linux
// runners resolvable instead of panicking.
#[cfg(not(any(target_os = "windows", target_os = "macos")))]
fn platform_data_dir() -> PathBuf {
    std::env::var_os("XDG_DATA_HOME")
        .map(PathBuf::from)
        .or_else(|| std::env::var_os("HOME").map(|home| PathBuf::from(home).join(".local/share")))
        .unwrap_or_else(|| PathBuf::from("."))
}

/// Read consent from `path`. A missing file or one that fails to parse
/// (corrupt JSON, or an unrecognized `consent` value) is treated as `Unset`
/// (fail closed, no telemetry) -- never an error.
pub fn read_consent(path: &Path) -> Consent {
    let Ok(data) = fs::read_to_string(path) else {
        return Consent::Unset;
    };
    serde_json::from_str::<StoredConsent>(&data)
        .map(|s| s.consent)
        .unwrap_or(Consent::Unset)
}

/// Persist `consent` to `path` as JSON.
pub fn write_consent(path: &Path, consent: Consent) -> Result<(), String> {
    if let Some(parent) = path.parent() {
        fs::create_dir_all(parent).map_err(|e| e.to_string())?;
    }
    let json =
        serde_json::to_string_pretty(&StoredConsent { consent }).map_err(|e| e.to_string())?;
    fs::write(path, json).map_err(|e| e.to_string())
}

/// Path segments that mark the project/app root in a stack frame -- when one
/// of these appears in a path, everything before it is stripped rather than
/// just redacting the home directory.
fn is_project_marker(segment: &str) -> bool {
    matches!(segment, "agentpanel" | "AgentPanel" | "src-tauri" | "src")
}

/// Matches an absolute path rooted at any recognized home/mount prefix:
/// macOS/Unix `/Users/name/...`, Windows `C:\Users\name\...` or
/// `C:/Users/name/...` (any drive letter case, either separator),
/// Linux `/home/name/...` and `/root` (no separate username segment -- it's
/// the root account's own home dir), macOS volume mounts `/Volumes/name/...`,
/// and Windows UNC shares `\\server\share\...`. Captures the whole
/// whitespace-delimited path token so trailing `:line:col` survives redaction.
fn home_path_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| {
        Regex::new(
            r"(?:/Users/|/home/|/Volumes/|/root\b|[A-Za-z]:[\\/]Users[\\/]|\\\\[^\\/\s]+\\[^\\/\s]+)[^\s]*",
        )
        .unwrap()
    })
}

/// Safety-net check used after the main redaction pass: if any of the
/// recognized home/mount prefixes still survives (any drive-letter case,
/// either Windows separator), the text couldn't be scrubbed confidently.
fn unscrubbed_marker_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| {
        Regex::new(
            r"(?:/Users/|/home/|/Volumes/|/root\b|[A-Za-z]:[\\/]Users[\\/]|\\\\[^\\/\s]+\\[^\\/\s]+)",
        )
        .unwrap()
    })
}

/// How many leading path segments make up the home/mount prefix to fold into
/// `~`, given the token already split on its separator. Two-segment prefixes
/// (`Users`/`home`/`Volumes` + the following name segment) fold both; `root`
/// is a one-segment prefix (`/root` **is** the home dir, no name follows);
/// UNC shares fold the leading `("", "", server, share)` run.
fn home_prefix_fold_len(segments: &[&str]) -> Option<usize> {
    if let Some(idx) = segments.iter().position(|s| matches!(*s, "Users" | "home" | "Volumes")) {
        return Some((idx + 2).min(segments.len()));
    }
    if let Some(idx) = segments.iter().position(|s| *s == "root") {
        return Some(idx + 1);
    }
    if segments.len() >= 4 && segments[0].is_empty() && segments[1].is_empty() {
        // Leading "\\server\share" splits (on '\\') into ["", "", server, share, ...].
        return Some(4);
    }
    None
}

/// Redact a single matched path token: strip everything before a project-root
/// marker segment when one is present; otherwise fall back to home/mount
/// redaction (replace the recognized prefix with `~`).
fn redact_path_token(token: &str) -> String {
    let sep = if token.contains('\\') { '\\' } else { '/' };
    let sep_str = sep.to_string();
    let segments: Vec<&str> = token.split(sep).collect();

    if let Some(marker_idx) = segments.iter().position(|s| is_project_marker(s)) {
        return segments[marker_idx..].join(&sep_str);
    }

    match home_prefix_fold_len(&segments) {
        Some(fold_len) => {
            let rest = &segments[fold_len..];
            if rest.is_empty() {
                "~".to_string()
            } else {
                format!("~{sep}{}", rest.join(&sep_str))
            }
        }
        None => token.to_string(),
    }
}

/// Redact home-directory/mount-point segments from Windows-, macOS- and
/// Linux-style absolute paths appearing anywhere in `input` (exception
/// values, messages, stack frames) -- see `home_path_re` for the recognized
/// prefixes. Non-path text is left intact. If, after redaction, one of those
/// prefixes still survives (i.e. the message can't be scrubbed with
/// confidence), the string is truncated at that point rather than sent with a
/// username/host attached.
pub fn scrub_paths(input: &str) -> String {
    let mut out = String::with_capacity(input.len());
    let mut last = 0;
    for m in home_path_re().find_iter(input) {
        out.push_str(&input[last..m.start()]);
        out.push_str(&redact_path_token(m.as_str()));
        last = m.end();
    }
    out.push_str(&input[last..]);

    if let Some(m) = unscrubbed_marker_re().find(&out) {
        out.truncate(m.start());
    }
    out
}

/// Response shape for the `get_telemetry_consent` command: the current file
/// value plus whether the Rust SDK actually initialized this session (the
/// value captured at startup, before this file could have changed).
#[derive(Debug, Clone, Serialize)]
pub struct TelemetryConsentInfo {
    pub consent: Consent,
    pub active_this_session: bool,
}

#[tauri::command]
pub fn get_telemetry_consent() -> TelemetryConsentInfo {
    TelemetryConsentInfo {
        consent: read_consent(&consent_path()),
        active_this_session: crate::telemetry_active_this_session(),
    }
}

/// Update the consent file. The frontend never touches the file directly;
/// this is the only write path. Takes effect on next launch -- it does not
/// retroactively start or stop reporting this session (see module docs).
#[tauri::command]
pub fn set_telemetry_consent(consent: Consent) -> Result<(), String> {
    write_consent(&consent_path(), consent)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use std::path::PathBuf;

    fn tmp_path(name: &str) -> PathBuf {
        std::env::temp_dir().join(format!(
            "agentpanel_telemetry_test_{}_{}.json",
            std::process::id(),
            name
        ))
    }

    // --- read_consent: file read/parse behavior -----------------------

    #[test]
    fn missing_file_reads_as_unset() {
        let path = tmp_path("missing");
        let _ = fs::remove_file(&path);
        assert_eq!(read_consent(&path), Consent::Unset);
    }

    #[test]
    fn corrupt_json_reads_as_unset() {
        let path = tmp_path("corrupt");
        fs::write(&path, "{ this is not valid json").unwrap();
        assert_eq!(read_consent(&path), Consent::Unset);
        let _ = fs::remove_file(&path);
    }

    #[test]
    fn unrecognized_consent_value_reads_as_unset() {
        let path = tmp_path("badvalue");
        fs::write(&path, r#"{"consent": "maybe"}"#).unwrap();
        assert_eq!(read_consent(&path), Consent::Unset);
        let _ = fs::remove_file(&path);
    }

    #[test]
    fn explicit_unset_value_reads_as_unset() {
        let path = tmp_path("explicit_unset");
        fs::write(&path, r#"{"consent": "unset"}"#).unwrap();
        assert_eq!(read_consent(&path), Consent::Unset);
        let _ = fs::remove_file(&path);
    }

    // --- write_consent / read_consent round trip -----------------------

    #[test]
    fn round_trip_granted() {
        let path = tmp_path("granted");
        write_consent(&path, Consent::Granted).unwrap();
        assert_eq!(read_consent(&path), Consent::Granted);
        let _ = fs::remove_file(&path);
    }

    #[test]
    fn round_trip_denied() {
        let path = tmp_path("denied");
        write_consent(&path, Consent::Denied).unwrap();
        assert_eq!(read_consent(&path), Consent::Denied);
        let _ = fs::remove_file(&path);
    }

    #[test]
    fn round_trip_unset() {
        let path = tmp_path("unset_roundtrip");
        write_consent(&path, Consent::Unset).unwrap();
        assert_eq!(read_consent(&path), Consent::Unset);
        let _ = fs::remove_file(&path);
    }

    #[test]
    fn write_then_overwrite_reflects_latest_value() {
        let path = tmp_path("overwrite");
        write_consent(&path, Consent::Granted).unwrap();
        assert_eq!(read_consent(&path), Consent::Granted);
        write_consent(&path, Consent::Denied).unwrap();
        assert_eq!(read_consent(&path), Consent::Denied);
        let _ = fs::remove_file(&path);
    }

    // --- scrub_paths: home-directory / username redaction ---------------
    //
    // The spec does not pin down the exact replacement token (e.g. "~",
    // "<home>", or something else) used when a path is scrubbed -- only that
    // the username/home directory must not survive, and that messages are
    // "kept but path-scrubbed" (not dropped). These tests assert that
    // invariant rather than an exact scrubbed string, since asserting a
    // specific token would be guessing at an unstated resolution.

    #[test]
    fn scrub_paths_redacts_macos_username() {
        let input = "Error at /Users/jane/dev/agentpanel/src-tauri/src/main.rs:42";
        let out = scrub_paths(input);
        assert!(!out.contains("jane"), "username should be redacted: {out}");
        assert!(!out.contains("/Users/jane"), "home dir should be redacted: {out}");
    }

    #[test]
    fn scrub_paths_redacts_windows_username() {
        let input = r"Error at C:\Users\jane\dev\agentpanel\src-tauri\src\main.rs:42";
        let out = scrub_paths(input);
        assert!(!out.contains("jane"), "username should be redacted: {out}");
        assert!(
            !out.contains(r"C:\Users\jane"),
            "home dir should be redacted: {out}"
        );
    }

    #[test]
    fn scrub_paths_keeps_message_text_intact() {
        let input = "panic: index out of bounds for /Users/jane/repo/src/lib.rs";
        let out = scrub_paths(input);
        assert!(
            out.contains("panic: index out of bounds"),
            "non-path message text should be preserved: {out}"
        );
        assert!(!out.contains("jane"));
    }

    // --- scrub_paths: additional home/mount prefixes (fix round h) ------

    #[test]
    fn scrub_paths_redacts_linux_home_username() {
        let input = "Error at /home/jane/dev/agentpanel/src/lib.rs:10";
        let out = scrub_paths(input);
        assert!(!out.contains("jane"), "username should be redacted: {out}");
        assert!(!out.contains("/home/jane"), "home dir should be redacted: {out}");
    }

    #[test]
    fn scrub_paths_redacts_root_home() {
        let input = "Error at /root/repo/src/lib.rs:10";
        let out = scrub_paths(input);
        assert!(!out.contains("/root"), "root home dir should be redacted: {out}");
    }

    #[test]
    fn scrub_paths_redacts_macos_volume_mount() {
        let input = "Error at /Volumes/Untitled/dev/agentpanel/src/lib.rs:10";
        let out = scrub_paths(input);
        assert!(!out.contains("Untitled"), "volume name should be redacted: {out}");
        assert!(!out.contains("/Volumes/Untitled"), "mount prefix should be redacted: {out}");
    }

    #[test]
    fn scrub_paths_redacts_unc_share() {
        let input = r"Error at \\build-server\repos\agentpanel\src\lib.rs:10";
        let out = scrub_paths(input);
        assert!(!out.contains("build-server"), "server name should be redacted: {out}");
        assert!(!out.contains("repos"), "share name should be redacted: {out}");
    }

    #[test]
    fn scrub_paths_redacts_lowercase_drive_forward_slash_windows_path() {
        // Lowercase drive letter + forward-slash separator (Git-Bash-style
        // paths); regression coverage for the widened truncation backstop,
        // which previously only matched uppercase `[A-Z]:\Users\`.
        let input = "Error at c:/Users/jane/dev/agentpanel/src/lib.rs:10";
        let out = scrub_paths(input);
        assert!(!out.contains("jane"), "username should be redacted: {out}");
    }
}
